~ Null handled incorrectly for *string

When decoding into a *string, null is handled correctly, but ~ is not.

To confirm this, I added the following two tests to `unmarshalTests` in decode_test.go:

``` go
{
    "canonical: ~",
    map[string]*string{"canonical": nil},
}, {
    "english: null",
    map[string]*string{"english": nil},
},
```

The first test fails, but the second test passes. My understanding was that `~` and `null` are semantically equivalent in yaml.

I tried again on a different machine, and still go the same result. I'm using a freshly-pulled version of the repository on master (7ad95dd0798a40da1ccdff6dff35fd177b5edf40) with go 1.4.2. My bash output and patch file are shown below, if they help.

One thing that I find strange is that even after I added 6 new tests (to see whether this is just a problem with `*string`, or pointers in general), the output from `go test` still only says it ran 21 tests. Does the test runner only run the first 21 tests for some reason?

``` bash
zachary $ go version
go version go1.4.2 linux/amd64
zachary $ export GOPATH=~/yaml-test/
zachary $ go get gopkg.in/yaml.v2
zachary $ go get gopkg.in/check.v1
zachary $ cd ~/yaml-test/src/gopkg.in/yaml.v2
zachary $ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
zachary $ git show --name-only
commit 7ad95dd0798a40da1ccdff6dff35fd177b5edf40
Author: Gustavo Niemeyer <gustavo@niemeyer.net>
Date:   Wed Jun 24 11:29:02 2015 +0100

    Fix omitempty support for floats.

encode_test.go
yaml.go
zachary $ go test
OK: 21 passed
PASS
ok      gopkg.in/yaml.v2    0.015s
zachary $ patch decode_test.go < ~/decode_test.patch 
patching file decode_test.go
zachary $ go test

----------------------------------------------------------------------
FAIL: decode_test.go:596: S.TestUnmarshal

decode_test.go:617:
    c.Assert(value, DeepEquals, item.value)
... obtained map[string]*string = map[string]*string{"canonical":(*string)(0xc2080a2140)}
... expected map[string]*string = map[string]*string{"canonical":(*string)(nil)}

OOPS: 20 passed, 1 FAILED
--- FAIL: Test (0.01s)
FAIL
exit status 1
FAIL    gopkg.in/yaml.v2    0.014s
zachary $ 
```

``` diff
--- gopkg.in/yaml.v2/decode_test.go 2015-06-25 09:33:53.192381256 -0700
+++ decode_test.go  2015-06-25 09:42:41.837811088 -0700
@@ -152,6 +152,27 @@
        map[string]*bool{"empty": nil},
    },

+   // For issue 109
+   {
+       "canonical: ~",
+       map[string]*string{"canonical": nil},
+   }, {
+       "english: null",
+       map[string]*string{"english": nil},
+   }, {
+       "canonical: ~",
+       map[string]*int{"canonical": nil},
+   }, {
+       "english: null",
+       map[string]*int{"english": nil},
+   }, {
+       "canonical: ~",
+       map[string]*struct{}{"canonical": nil},
+   }, {
+       "english: null",
+       map[string]*struct{}{"english": nil},
+   },
+   
    // Flow sequence
    {
        "seq: [A,B]",
```

--- a/decode.go
+++ b/decode.go
@@ -251,7 +251,7 @@ func (d *decoder) callUnmarshaler(n *node, u Unmarshaler) (good bool) {
 //
 // If n holds a null value, prepare returns before doing anything.
 func (d *decoder) prepare(n *node, out reflect.Value) (newout reflect.Value, unmarshaled, good bool) {
-	if n.tag == yaml_NULL_TAG || n.kind == scalarNode && n.tag == "" && (n.value == "null" || n.value == "" && n.implicit) {
+	if n.tag == yaml_NULL_TAG || n.kind == scalarNode && n.tag == "" && (n.value == "null" || n.value == "~" || n.value == "" && n.implicit) {
 		return out, false, false
 	}
 	again := true

--- a/decode_test.go
+++ b/decode_test.go
@@ -440,9 +440,24 @@ var unmarshalTests = []struct {
 		map[string]*string{"foo": new(string)},
 	}, {
 		"foo: null",
+		map[string]*string{"foo": nil},
+	}, {
+		"foo: null",
 		map[string]string{"foo": ""},
 	}, {
 		"foo: null",
+		map[string]interface{}{"foo": nil},
+	},
+
+	// Support for ~
+	{
+		"foo: ~",
+		map[string]*string{"foo": nil},
+	}, {
+		"foo: ~",
+		map[string]string{"foo": ""},
+	}, {
+		"foo: ~",
 		map[string]interface{}{"foo": nil},
 	},
 

